File: apps/windows/ClipPlus.Windows/Sync/ClipPlusMessage.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClipPlus.Windows.Sync;

public enum ClipPlusMessageKind
{
    Hello,
    Trust,
    Text,
    Image,
    ImageOffer,
    FileOffer
}

public sealed record FileTransferItem(string RelativePath, long ByteSize, bool IsDirectory);

public enum FileTransferFormat
{
    DirectTree
}

public sealed class ClipPlusMessage
{
    public const int MaxInlineImageBytes = 32 * 1024;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) }
    };

    public ClipPlusMessageKind Kind { get; init; }
    public int ProtocolVersion { get; init; } = 1;
    public string GroupId { get; init; } = string.Empty;
    public string SenderDeviceId { get; init; } = string.Empty;
    public string SenderDeviceName { get; init; } = string.Empty;
    public string EventId { get; init; } = Guid.NewGuid().ToString();
    public string? Text { get; init; }
    public string? ImageBase64 { get; init; }
    public int? ImageByteSize { get; init; }
    public string? ImageContentHash { get; init; }
    public string? ApprovedDeviceId { get; init; }
    public string? TransferId { get; init; }
    public FileTransferFormat? TransferFormat { get; init; }
    public IReadOnlyList<FileTransferItem>? Files { get; init; }
    public int? ArchivePort { get; init; }
    public string CreatedAt { get; init; } = DateTimeOffset.UtcNow.ToString("O");

    [JsonIgnore]
    public byte[]? DecodedImageData => ImageBase64 is null ? null : Convert.FromBase64String(ImageBase64);

    public static ClipPlusMessage CreateHello(
        string groupId,
        string senderDeviceId,
        string senderDeviceName)
    {
        var json = new ClipPlus.Windows.CoreBridge.CoreBridge().CreateHelloMessageJson(
            groupId,
            senderDeviceId,
            senderDeviceName
        ) ?? throw new InvalidOperationException("Rust core library is unavailable; cannot create hello message.");

        return FromJson(json);
    }

    public static ClipPlusMessage CreateTrust(
        string groupId,
        string senderDeviceId,
        string senderDeviceName,
        string approvedDeviceId)
    {
        var json = new ClipPlus.Windows.CoreBridge.CoreBridge().CreateTrustMessageJson(
            groupId,
            senderDeviceId,
            senderDeviceName,
            approvedDeviceId
        ) ?? throw new InvalidOperationException("Rust core library is unavailable; cannot create trust message.");

        return FromJson(json);
    }

    public static ClipPlusMessage CreateFileOffer(
        string groupId,
        string senderDeviceId,
        string senderDeviceName,
        string transferId,
        IReadOnlyList<FileTransferItem> files,
        int archivePort)
    {
        var json = new ClipPlus.Windows.CoreBridge.CoreBridge().CreateFileOfferMessageJson(
            groupId,
            senderDeviceId,
            senderDeviceName,
            transferId,
            files,
            archivePort
        ) ?? throw new InvalidOperationException("Rust core library is unavailable; cannot create file offer message.");

        return FromJson(json);
    }

    public static ClipPlusMessage CreateImageOffer(
        string groupId,
        string senderDeviceId,
        string senderDeviceName,
        string transferId,
        byte[] pngData,
        int archivePort)
    {
        var json = new ClipPlus.Windows.CoreBridge.CoreBridge().CreateImageOfferMessageJson(
            groupId,
            senderDeviceId,
            senderDeviceName,
            transferId,
            pngData,
            archivePort
        ) ?? throw new InvalidOperationException("Rust core library is unavailable; cannot create image offer message.");

        return FromJson(json);
    }

    public static ClipPlusMessage? CreateImage(
        string groupId,
        string senderDeviceId,
        string senderDeviceName,
        byte[] pngData)
    {
        if (pngData.Length == 0 || pngData.Length > MaxInlineImageBytes)
        {
            return null;
        }

        var json = new ClipPlus.Windows.CoreBridge.CoreBridge().CreateImageMessageJson(
            groupId,
            senderDeviceId,
            senderDeviceName,
            pngData
        ) ?? throw new InvalidOperationException("Rust core library is unavailable; cannot create image clipboard message.");

        return FromJson(json);
    }

    public static ClipPlusMessage CreateText(
        string groupId,
        string senderDeviceId,
        string senderDeviceName,
        string text)
    {
        var json = new ClipPlus.Windows.CoreBridge.CoreBridge().CreateTextMessageJson(
            groupId,
            senderDeviceId,
            senderDeviceName,
            text
        ) ?? throw new InvalidOperationException("Rust core library is unavailable; cannot create text clipboard message.");

        return FromJson(json);
    }

    public string ToJson()
    {
        return JsonSerializer.Serialize(this, JsonOptions);
    }

    public static ClipPlusMessage FromJson(string json)
    {
        return JsonSerializer.Deserialize<ClipPlusMessage>(json, JsonOptions)
            ?? throw new InvalidOperationException("ClipPlus message JSON is empty");
    }
}

File: apps/mac/Sources/ClipPlusMac/Sync/ClipPlusMessage.swift
import Foundation

enum ClipPlusMessageKind: String, Codable {
    case hello
    case trust
    case text
    case image
    case imageOffer
    case fileOffer
}

struct FileTransferItem: Codable, Equatable {
    let relativePath: String
    let byteSize: Int64
    let isDirectory: Bool
}

enum FileTransferFormat: String, Codable {
    case directTree
}

struct ClipPlusMessage: Codable, Equatable {
    static let maxInlineImageBytes = 32 * 1024

    let kind: ClipPlusMessageKind
    let protocolVersion: Int
    let groupId: String
    let senderDeviceId: String
    let senderDeviceName: String
    let eventId: String
    let text: String?
    let imageBase64: String?
    let imageByteSize: Int?
    let imageContentHash: String?
    let approvedDeviceId: String?
    let transferId: String?
    let transferFormat: FileTransferFormat?
    let files: [FileTransferItem]?
    let archivePort: Int?
    let createdAt: String

    var decodedImageData: Data? {
        guard let imageBase64 else {
            return nil
        }

        return Data(base64Encoded: imageBase64)
    }

    static func hello(
        groupId: String,
        senderDeviceId: String,
        senderDeviceName: String
    ) -> Self? {
        guard let json = CoreBridge().createHelloMessageJSON(
            groupId: groupId,
            senderDeviceId: senderDeviceId,
            senderDeviceName: senderDeviceName
        ) else {
            return nil
        }

        return decodeCoreMessageJSON(json)
    }

    static func trust(
        groupId: String,
        senderDeviceId: String,
        senderDeviceName: String,
        approvedDeviceId: String
    ) -> Self? {
        guard let json = CoreBridge().createTrustMessageJSON(
            groupId: groupId,
            senderDeviceId: senderDeviceId,
            senderDeviceName: senderDeviceName,
            approvedDeviceId: approvedDeviceId
        ) else {
            return nil
        }

        return decodeCoreMessageJSON(json)
    }

    static func text(
        groupId: String,
        senderDeviceId: String,
        senderDeviceName: String,
        text: String
    ) -> Self? {
        guard let json = CoreBridge().createTextMessageJSON(
            groupId: groupId,
            senderDeviceId: senderDeviceId,
            senderDeviceName: senderDeviceName,
            text: text
        ) else {
            return nil
        }

        return decodeCoreMessageJSON(json)
    }

    static func image(
        groupId: String,
        senderDeviceId: String,
        senderDeviceName: String,
        pngData: Data
    ) -> Self? {
        guard !pngData.isEmpty, pngData.count <= maxInlineImageBytes else {
            return nil
        }

        guard let json = CoreBridge().createImageMessageJSON(
            groupId: groupId,
            senderDeviceId: senderDeviceId,
            senderDeviceName: senderDeviceName,
            pngData: pngData
        ) else {
            return nil
        }

        return decodeCoreMessageJSON(json)
    }

    static func fileOffer(
        groupId: String,
        senderDeviceId: String,
        senderDeviceName: String,
        transferId: String,
        files: [FileTransferItem],
        archivePort: Int
    ) -> Self? {
        guard let json = CoreBridge().createFileOfferMessageJSON(
            groupId: groupId,
            senderDeviceId: senderDeviceId,
            senderDeviceName: senderDeviceName,
            transferId: transferId,
            files: files,
            archivePort: archivePort
        ) else {
            return nil
        }

        return decodeCoreMessageJSON(json)
    }

    static func imageOffer(
        groupId: String,
        senderDeviceId: String,
        senderDeviceName: String,
        transferId: String,
        pngData: Data,
        archivePort: Int
    ) -> Self? {
        guard let json = CoreBridge().createImageOfferMessageJSON(
            groupId: groupId,
            senderDeviceId: senderDeviceId,
            senderDeviceName: senderDeviceName,
            transferId: transferId,
            pngData: pngData,
            archivePort: archivePort
        ) else {
            return nil
        }

        return decodeCoreMessageJSON(json)
    }

    private static func decodeCoreMessageJSON(_ json: String) -> Self? {
        guard let data = json.data(using: .utf8),
              let message = try? JSONDecoder().decode(Self.self, from: data) else {
            return nil
        }

        return message
    }
}
